corehq/apps/app_manager/static/app_manager/js/app_manager.js

```diff
@@ -518,23 +532,136 @@ hqDefine('app_manager/js/app_manager', [
     });
 
     var _initNewModuleOptionClicks = function () {
+        var self = {};
+        self.deprecatedCaseTypes = ko.observableArray();
+
         $('.new-module-option').on('click', function () {
             var moduleType = $(this).data('type');
             $('#new-module-type').val(moduleType);
             var $form = $('#new-module-form');
-            $('.new-module-icon').removeClass().addClass("fa fa-refresh fa-spin");
+
             if (moduleType === "case") {
-                google.track.event("Added Case List Menu");
-                kissmetrix.track.event("Added Case List Menu");
-            } else if (moduleType === "survey") {
-                google.track.event("Added Surveys Menu");
-                kissmetrix.track.event("Added Surveys Menu");
+                $('#add-new-module-modal').modal('hide');
+                $('#define-case-type-modal').modal('show');
+            } else {
+                if (moduleType === "survey") {
+                    google.track.event("Added Surveys Menu");
+                    kissmetrix.track.event("Added Surveys Menu");
+                }
+                $('.new-module-icon').removeClass().addClass("fa fa-refresh fa-spin");
+                $('#add-new-module-modal').modal('hide');
+                $form.submit();
+            }
+        });
+
+        $('#define-case-type-modal').on('show.bs.modal', function () {
+            var $caseType = $('#new_case_type');
+            if (!$caseType.data('select2')) {
+                $caseType.select2({
+                    tags: true,
+                    placeholder: gettext("Add a case type or look for an existing one"),
+                    allowClear: true,
+                    width: '100%',
+                    createTag: function (params) {
+                        // Replace spaces with underscore
+                        var term = params.term.replace(/ /g, '_');
+                        return {
+                            id: term,
+                            text: term,
+                        };
+                    },
+                });
+
+                $.ajax({
+                    method: 'GET',
+                    url: initialPageData.reverse('all_case_types'),
+                    success: function (data) {
+                        var existingCaseTypes = data.existing_case_types;
+                        self.deprecatedCaseTypes(data.deprecated_case_types);
+
+                        // Add existing case types as options
+                        existingCaseTypes.forEach(function (caseType) {
+                            var option = new Option(caseType, caseType, false, false);
+                            $caseType.append(option);
+                        });
+                    },
+                });
```

Comment:
Can you add a comment clarifying the use of the ajax block here rather than in the select2 constructor?

Reply:
13ae7b64894a3d6d0a6ca66f40a98d08adaeb6a5